Added new option "use_output_inserted_into" that defaults to false.

saeohioalpha opened the pull request:
Added new insert condition to use the OUTPUT INSERTED INTO TSQL statement format when the use_output_inserted_into option is set to true.
If the option is false, it uses the existing TSQL statement

This is due to SQL Server requiring the OUTPUT INSERTED INTO when a DB Table has a trigger. In that case a simple OUTPUT INSERTED will not work and will throw an error.

> The target table 'T' of the DML statement cannot have any enabled triggers if the statement contains an OUTPUT clause without INTO clause.

#593

metaskills commented:
Thanks, I finally got the Rails v5.1 work done. In your core issue, you stated the following:

> On a table with a trigger, you have to do a OUTPUT INSERTED.id INTO not just OUTPUT INSERTED.id

Why not use `OUTPUT INSERTED.id INTO` all the time vs adding another config?

metaskills commented:
```
DECLARE @inserted_id table (id int)
```

This looks like it has a few issues. First, it would be the `quoted_pk`, second bigint is the default primary key for Rails now and we want to cast (like we do elsewhere) in this method to that type. We can't assume int for everything.

metaskills commented on lib/active_record/connection_adapters/sqlserver_adapter.rb:
```diff
@@ -44,6 +44,7 @@ class SQLServerAdapter < AbstractAdapter
 
       cattr_accessor :cs_equality_operator, instance_accessor: false
       cattr_accessor :use_output_inserted, instance_accessor: false
+      cattr_accessor :use_output_inserted_into, instance_accessor: false
```
Would like to avoid using this and assume `use_output_inserted` leverages INTO as our default now

metaskills replied:
Would like to avoid using this and assume use_output_inserted leverages INTO as our default now

metaskills commented on lib/active_record/connection_adapters/sqlserver/database_statements.rb:
```diff
@@ -198,7 +198,12 @@ def sql_for_insert(sql, pk, id_value, sequence_name, binds)
+                    sql.dup.insert sql.index(/ (DEFAULT )?VALUES/), " OUTPUT INSERTED.#{quoted_pk}"
+                  else
+                    tempsql = sql.insert sql.index(/ (DEFAULT )?VALUES/), " OUTPUT INSERTED.#{quoted_pk} INTO @inserted_id"
+                    "DECLARE @inserted_id table (id int);#{tempsql}; SELECT * FROM @inserted_id;"
```
* Assume `use_output_inserted` uses `INTO`. No need for two formats.
* Change `(id int)` to something like `CAST(#{quoted_pk} AS bigint)`

saeohioalpha commented:
That could definitely be done. But since TSQL supported both I figured there may be use cases I am unaware of when someone would not want to use the updated language and I made it with that idea.

Sent from my iPhone

> On May 28, 2017, at 3:49 PM, Ken Collins <notifications@github.com> wrote:
> 
> DECLARE @inserted_id table (id int)
> This looks like it has a few issues. First, it would be the quoted_pk, second bigint is the default primary key for Rails now and we want to cast (like we do elsewhere) in this method to that type. We can't assume int for everything.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

saeohioalpha commented:
I removed the extra option for :use_output_inserted_into
I have combined the conditionals to use the INTO format.
Changed the identity to use bigint as its datatype.

metaskills commented:
Thanks @saeohioalpha, are you actively working on this to get the tests passing? Right now it looks to be bombing out on the first fixture insert.

sardaukar commented:
Is there a workaround for this problem for the time being?

saeohioalpha commented:
Whoops, just saw this needed attention, I thought it had been merged in. Ill take a look.

saeohioalpha commented:
So im looking at the appveyor error and its crashing on the initial migration. 

`TinyTds::Error: Error converting data type nvarchar to bigint.: EXEC sp_executesql N'DECLARE @inserted_identity table ([key] bigint);INSERT INTO [ar_internal_metadata] ([key], [value], [created_at], [updated_at]) OUTPUT INSERTED.[key] INTO @inserted_identity VALUES (@0, @1, @2, @3);SELECT CAST([key] AS bigint) FROM @inserted_identity;', N'@0 nvarchar(4000), @1 nvarchar(4000), @2 datetime, @3 datetime', @0 = N'environment', @1 = N'development', @2 = '06-05-2017 20:55:46.333', @3 = '06-05-2017 20:55:46.333'`

It looks like this (ar_internal_metadata) is the table rails stores the migration information in. I have no idea why activerecord-sqlserver-adapter is trying to get the identity for a key->value table. Shouldn't it not be doing an identity insert on this table?

metaskills commented:
Closing this in favor of #625. A few things that drove that implementation. 

* The test here failed for any table that did not use a biginit PK.
* The overhead for the tmp table should be opt in.
* The tmp table output inserted into should be per table configurable.

metaskills commented:
Released v5.1.2 with this and another fix.
